File: android/src/main/kotlin/com/meg4cyberc4t/varioqub_configs/VarioqubConfigsPlugin.kt
package com.meg4cyberc4t.varioqub_configs

import PigeonBuildSettings
import VarioqubSender
import android.content.Context
import android.util.Log
import com.yandex.varioqub.config.FetchError
import com.yandex.varioqub.config.OnFetchCompleteListener
import com.yandex.varioqub.config.Varioqub
import com.yandex.varioqub.config.VarioqubApi
import com.yandex.varioqub.config.VarioqubSettings

import io.flutter.embedding.engine.plugins.FlutterPlugin


/** VarioqubConfigsPlugin */
class VarioqubConfigsPlugin : FlutterPlugin, VarioqubSender {
    private companion object {
        private const val TAG = "VarioqubConfigsPlugin"
    }

    private lateinit var api: VarioqubApi
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        try {
            VarioqubSender.setUp(flutterPluginBinding.binaryMessenger, this)
            context = flutterPluginBinding.applicationContext
            api = Varioqub.getInstance()
        } catch (error: Throwable) {
            Log.e(TAG, "Failed to initialize TickerPlugin")
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        VarioqubSender.setUp(binding.binaryMessenger, null)
        Varioqub.activateConfig()
    }

    override fun build(settings: PigeonBuildSettings) {
        var settingsBuilder =
            VarioqubSettings.Builder(settings.clientId)
                .withActivateEvent(settings.activateEvent)
        if (settings.logs == true) {
            settingsBuilder.withLogs().also { settingsBuilder = it };
        }
        settings.clientFeatures.forEach {
            if (it.key != null && it.value != null) {
                settingsBuilder = settingsBuilder.withClientFeature(it.key!!, it.value!!)
            }
        }
        api.init(
            settingsBuilder.build(),
            VarioqubAdapter(),
            context,
        );
    }

    override fun fetchConfig(callback: (Result<Unit>) -> Unit) {
        api.fetchConfig(
            object : OnFetchCompleteListener {
                override fun onSuccess() {
                    callback(Result.success(Unit))
                }

                override fun onError(message: String, error: FetchError) {
                    Log.e(TAG, message);
                    Log.e(TAG, error.toString());
                    callback(Result.failure(Throwable(message)))
                }
            }

        )
    }

    override fun activateConfig(callback: (Result<Unit>) -> Unit) {
        try {
            api.activateConfig(
                onComplete = {
                    callback(Result.success(Unit))
                }
            )
        } catch (exception: Throwable) {
            callback(Result.failure(exception))
        }
    }

    override fun setDefaults(values: Map<String, Any>) {
        return api.setDefaults(values);
    }

    override fun getString(key: String, defaultValue: String): String {
        return api.getString(key = key, default = defaultValue)
    }

    override fun getBool(key: String, defaultValue: Boolean): Boolean {
        return api.getBoolean(key = key, default = defaultValue)
    }

    override fun getInt(key: String, defaultValue: Long): Long {
        return api.getLong(key = key, default = defaultValue)
    }

    override fun getDouble(key: String, defaultValue: Double): Double {
        return api.getDouble(key = key, default = defaultValue)
    }

    override fun getId(): String {
        return api.getId()
    }

    override fun putClientFeature(key: String, value: String) {
        return api.putClientFeature(key = key, value = value)
    }

    override fun clearClientFeatures() {
        return api.clearClientFeatures()
    }

    override fun getAllKeys(): List<String> {
        return api.getAllKeys().toList();
    }
}

File: ios/Classes/VarioqubConfigsPlugin.swift
//
//  VarioqubNullAdapter.swift
//  varioqub_configs
//
//  Created by Igor Molchanov on 27.02.2024.
//

import Varioqub
import MetricaAdapter

#if os(iOS)
import Flutter
#elseif os(macOS)
import FlutterMacOS
#else
#error("Unsupported platform.")
#endif



public class VarioqubApiPigeonImpl: NSObject, FlutterPlugin, VarioqubApiPigeon {
    
    
    var idProvider: VarioqubIdProvider? = nil;
    var reporter: VarioqubReporter? = nil;
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        super.init()
        VarioqubApiPigeonSetup.setUp(binaryMessenger: binaryMessenger, api: self)
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "varioqub_configs", binaryMessenger: registrar.messenger())
        let instance = VarioqubApiPigeonImpl(binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    func build(settings: VarioqubSettingsPigeon) throws {
        var config = VarioqubConfig.default
        if (settings.fetchThrottleIntervalMs != nil) {
            config.fetchThrottle = Double(settings.fetchThrottleIntervalMs! / 1000)
        }
        if (settings.url != nil) {
            config.baseURL = URL(string: settings.url!)
        }
        
        var clientFeatures: [String : String] = [:]
        for (key, value) in settings.clientFeatures {
            if let key = key, let value = value {
                clientFeatures[key] = value
            }
        }
        config.initialClientFeatures = ClientFeatures(dictionary: clientFeatures)
        
        
        VarioqubFacade.shared.initialize(
            clientId: settings.clientId,
            config: config,
            idProvider:  settings.trackingWithAppMetrica ? AppmetricaAdapter() :  VarioqubNullHandler(),
            reporter: settings.trackingWithAppMetrica ? AppmetricaAdapter() :  VarioqubNullHandler()
        )
    }
    
    
    func fetchConfig(completion: @escaping (Result<FetchErrorPigeon?, any Error>) -> Void) {
        VarioqubFacade.shared.fetchConfig({ status in
            switch (status) {
            case .success, .cached:
                completion(.success(nil))
                return
            case .throttled:
                completion(.success(FetchErrorPigeon(error: VarioqubFetchErrorPigeon.requestThrottled)))
                return
            case .error(let error):
                let message: String?
                let fetchError: VarioqubFetchErrorPigeon
                switch (error) {
                case .emptyResult:
                    fetchError = VarioqubFetchErrorPigeon.emptyResult
                    message = nil
                case .nullIdentifiers:
                    fetchError = VarioqubFetchErrorPigeon.identifiersNull
                    message = nil
                case .response(let error), .parse(let error):
                    fetchError = VarioqubFetchErrorPigeon.responseParseError
                    message = error.localizedDescription
                case .request:
                    fetchError = VarioqubFetchErrorPigeon.internalError
                    message = nil
                case .underlying(let error):
                    fetchError = VarioqubFetchErrorPigeon.internalError
                    message = error.localizedDescription
                case .network(let error):
                    fetchError = VarioqubFetchErrorPigeon.networkError
                    message = error.localizedDescription
                @unknown default:
                    fetchError = VarioqubFetchErrorPigeon.internalError
                    message = nil
                }
                completion(.success(FetchErrorPigeon(message: message, error: fetchError)))
                return
            @unknown default:
                completion(.success(nil))
                return
            }
        })
    }
    
    func activateConfig(completion: @escaping (Result<Void, Error>) -> Void) {
        VarioqubFacade.shared.activateConfigAndWait()
        completion(.success(Void()))
    }
    
    func setDefaults(values: [String : Any]) throws {
        var defaults = [VarioqubFlag: String]()
        for (key, value) in values {
            defaults[VarioqubFlag(rawValue: key)] = String(describing: value)
        }
        return VarioqubFacade.shared.setDefaultsAndWait(defaults)
    }
    
    func getString(key: String, defaultValue: String) throws -> String {
        return VarioqubFacade.shared.getString(for: VarioqubFlag(rawValue: key), defaultValue: defaultValue)
    }
    
    func getBool(key: String, defaultValue: Bool) throws -> Bool {
        return VarioqubFacade.shared.getBool(for: VarioqubFlag(rawValue: key), defaultValue: defaultValue)
    }
    
    func getInt(key: String, defaultValue: Int64) throws -> Int64 {
        return VarioqubFacade.shared.getInt64(for: VarioqubFlag(rawValue: key), defaultValue: defaultValue)
    }
    
    func getDouble(key: String, defaultValue: Double) throws -> Double {
        return VarioqubFacade.shared.getDouble(for: VarioqubFlag(rawValue: key), defaultValue: defaultValue)
    }
    
    func getId() throws -> String {
        return VarioqubFacade.shared.varioqubId ?? ""
    }
    
    func putClientFeature(key: String, value: String) throws {
        return VarioqubFacade.shared.clientFeatures.setFeature(value, forKey: key);
    }
    
    func clearClientFeatures() throws {
        return VarioqubFacade.shared.clientFeatures.clearFeatures()
    }
    
    func getAllKeys() throws -> [String] {
        return Array(VarioqubFacade.shared.allKeys.map {
            return $0.rawValue;
        })
    }
    
    func getAllValues() throws -> [String : String] {
        return VarioqubFacade.shared.allItems
            .map { key, value in (key.rawValue, value.stringValueOrDefault) }
            .reduce(into: [:]) { result, tuple in
                result[tuple.0] = tuple.1
            }
    }
}
